src/librustc_typeck/astconv.rs

```diff
@@ -673,45 +672,23 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                                         });
                                     });
-                                    let is_lifetime_present =
-                                        params_present.contains(&ParamKindOrd::Lifetime);
-                                    let is_types_present =
-                                        params_present.contains(&ParamKindOrd::Type);
-                                    let is_consts_present =
-                                        params_present.contains(&ParamKindOrd::Const);
 
                                     Self::generic_arg_mismatch_err(
                                         tcx.sess,
                                         arg,
                                         kind.descr(),
                                         Some(&format!(
-                                            "reorder the arguments: {}{}{}: `<{}>`",
-                                            if is_lifetime_present { "lifetimes" } else { "" },
-                                            if is_types_present {
-                                                if is_lifetime_present {
-                                                    ", then types"
-                                                } else {
-                                                    "types"
-                                                }
-                                            } else {
-                                                ""
-                                            },
-                                            if is_consts_present
-                                                && tcx.sess.features_untracked().const_generics
-                                            {
-                                                if is_lifetime_present || is_types_present {
-                                                    ", then consts"
-                                                } else {
-                                                    "consts"
-                                                }
-                                            } else {
-                                                ""
-                                            },
+                                            "reorder the arguments: {}: `<{}>`",
+                                            params_present
+                                                .into_iter()
+                                                .map(|param| param.to_string())
+                                                .collect::<Vec<String>>()
+                                                .join(", then "),
                                             ordered_params
                                                 .into_iter()
                                                 .filter_map(|param| {
                                                     let param_name = param.name.to_string();
```

Comment:
If you don't use `to_string()` here, you don't have to use it on `kw::SelfUpper`.